docs/GettingStartedDocs/best_practices_secure_enclaves.md

```diff
@@ -0,0 +1,242 @@
+1)	The pointer is not null.
+2)	The bounding arithmetic operations do not wrap (i.e. numerical overflow).
+3)	The block lies completely within or outside of the enclave.
+
+The first two conditions should make clear the pitfalls of considering the function pair to be boolean opposites: If the pointer is null or the bounding calculations overflow in a call to `oe_is_within_enclave()`, the function will return `false`, which should _not_ be interpreted as "the memory is outside the enclave". 
+
+The last condition is subtle: In some potentially malicious cases, the memory range may be both partially within and partially outside the enclave, spanning the boundary between the two. (Attacks against enclaves may use memory confusion to achieve overwrites of protected regions, for example.) Both functions will return `false` when the memory range spans the boundary - another reason the function pair are not boolean opposites.
+
+<br />
+<br />
+<br />
+
+## Time-of-check/Time-of-use or "double-fetch" vulnerabilities
+
+One class of vulnerabilities that your marshaling code should protect against is referred to as time-of-check/time-of-use, or TOCTOU. Another name for this problem is "double-fetch". The problem arises when memory is shared across privilege boundaries. As code validates input, it's critical that data is not fetched twice (or, strictly, more than once), allowing a malicious untrusted called to change the data between the time-of-check and the time-of-use. Let's examine our function:
+
+```c++
+// WARNING: Portions of this code are intentionally flawed to demonstrate common pitfalls.
+int ecall_with_user_check2(void* ptr) {
+
+    blob_t* blob;
+
+    // Ensure passed-in pointer is not null and the memory block is located entirely outside of the enclave.
+    if (!oe_is_outside_enclave(ptr, sizeof(blob_t)))
+        return -1;
+
+    blob = (blob_t*)ptr;
+
+    if (blob->size > 4096) //TOCTOU: First fetch
+        return -1;
+
+    return render_data(blob->data, blob->size); //TOCTOU: Second fetch
+}
+```
+There's nothing wrong with the first fetch: The value is validated, and the appropriate logic branch is taken. The problem arises when the code needs the value again and reads it, once again, from the untrusted memory location outside of the enclave. Between the first and subsequent fetch the value may have been changed, neutralizing the size validation and possibly leading to some enclave memory corruption that is helpful to the attacker. Let's protect the code by "capturing" the values.
+
+> When "capturing" data values in this context, it's important to protect against the compiler's optimization, hence we use the `volatile` qualifier for our local blob_t. Otherwise, the compiler might optimize-away our local variable, removing the TOCTOU protection.
```

Comment:
I don't think this application of volatile is correct. Furthermore, we have to consider which behavioral form (ISO, MS) is being used. 
I'm looking into this now.

Reply:
volatile belongs on blob, not captured_blob. that's the memory that could change at anytime.

Reply:
I think the current approach is okay, but I'm going change the approach a bit: instead of using the blob struct, I'm going switch to two volatile vars.